gdx/src/com/badlogic/gdx/graphics/g2d/PixmapPacker.java

```diff
@@ -326,66 +322,61 @@ public synchronized void dispose () {
-			if (page.texture == null) {
-				if (page.rects.size != 0 && page.addedRects.size > 0) {
-					page.texture = new Texture(new PixmapTextureData(page.image, page.image.getFormat(), useMipMaps, false, true)) {
-						@Override
-						public void dispose () {
-							super.dispose();
-							getTextureData().consumePixmap().dispose();
-						}
-					};
-					page.texture.setFilter(minFilter, magFilter);
-
-					for (String name : page.addedRects) {
-						Rectangle rect = page.rects.get(name);
-						TextureRegion region = new TextureRegion(page.texture, (int)rect.x, (int)rect.y, (int)rect.width,
-							(int)rect.height);
-						atlas.addRegion(name, region);
-					}
-					page.addedRects.clear();
-				}
-			} else {
-				if (page.addedRects.size > 0) {
+			if (page.addedRects.size > 0) {
+				if (page.texture == null) {
+					generatePageTexture(page, minFilter, magFilter, useMipMaps);
+				} else {
 					page.texture.load(page.texture.getTextureData());
-					for (String name : page.addedRects) {
-						Rectangle rect = page.rects.get(name);
-						TextureRegion region = new TextureRegion(page.texture, (int)rect.x, (int)rect.y, (int)rect.width,
+				}
+				for (String name : page.addedRects) {
+					Rectangle rect = page.rects.get(name);
+					TextureRegion region = new TextureRegion(page.texture, (int)rect.x, (int)rect.y, (int)rect.width,
 							(int)rect.height);
-						atlas.addRegion(name, region);
-					}
-					page.addedRects.clear();
+					atlas.addRegion(name, region);
 				}
+				page.addedRects.clear();
+			}
+		}
+	}
+
+	private void generatePageTexture(final Page page, TextureFilter minFilter, TextureFilter magFilter, boolean useMipMaps) {
+		page.texture = new Texture(new PixmapTextureData(page.image, page.image.getFormat(), useMipMaps, false, true)) {
+			@Override
+			public void dispose() {
+				super.dispose();
+				getTextureData().consumePixmap().dispose();
+			}
+		};
+		page.texture.setFilter(minFilter, magFilter);
+	}
+
+	/** Return an ordered array of texture regions corresponding with the pages of the packer, creating new textures
+	 * if necessary.  The same array will be returned from successive calls to this method, possibly modified with
+	 * additional texture regions if more pages have been added since the last call.  Once this method has been called,
+	 * you must dispose of the resulting texture regions (or equivalently, any atlas generated or updated by this packer)
+	 * yourself - disposing the PixmapPacker will no longer dispose these textures or their backing pixmaps.
+	 */
+	public Array<TextureRegion> getTextureRegions(TextureFilter minFilter, TextureFilter magFilter, boolean useMipMaps) {
+		while (textureRegions.size < pages.size) {
+			if (disposed) return null;
+			Page page = pages.get(textureRegions.size);
+			if (page.texture == null) {
+				generatePageTexture(page, minFilter, magFilter, useMipMaps);
 			}
+			textureRegions.add(new TextureRegion(page.texture));
 		}
+		return textureRegions;

```

Comment:
So the problem is if you call `updateTextureRegions` then call `updateTextureAtlas`, the atlas won't have any regions. Maybe the solution is to just not have two methods, delete `updateTextureRegions` and only have `updateTextureAtlas`. If you want the textures, use TextureAtlas `getTextures`.

For your third point, all up front packing that can be done needs to be done before creating the textures. That means `generateFont` should not be used, `generateFontData` should.

Here is some code that isn't how the API currently works, but it could:
```java
packer = new PixmapPacker();
// Use glTexSubImage2D instead of reloading whole texture. Required for incremental fonts.
packer.setIncremental(true);
fontGenerator = new FreeTypeFontGenerator(fontFile);
for (Pixmap pixmap : images1)
	packer.pack(name, pixmap);
fontData1 = new FreeTypeBitmapFontData(fontGenerator); // Adds glyphs to packer.
for (Pixmap pixmap : images2)
	packer.pack(name, pixmap);
fontData2 = new FreeTypeBitmapFontData(fontGenerator); // Adds glyphs to packer.
atlas = new TextureAtlas();
packer.updateTextureAtlas(atlas); // Generates textures for first time.
font1 = new BitmapFont(data1, atlas.getTextures());
font2 = new BitmapFont(data2, atlas.getTextures());
// ...
for (Pixmap pixmap : images3)
	packer.pack(name, pixmap); // Textures updated incrementally.
// ...
// FreeType font adds glyphs to packer, textures updated incrementally.
// ...
font2.dispose(); // Disposes textures, textures dispose pixmaps.
font1.dispose(); // Disposes same textures, nothing happens.
atlas.dispose(); // Disposes same textures, nothing happens.
fontGenerator.dispose();
packer.dispose(); // Disposes same pixmaps, nothing happens.
```

PixmapPacker gets a new setting which uses `glTexSubImage2D` when `pack` is called. In this case, `updateTextureAtlas` does not reload the whole texture, it only adds regions to the atlas.

Currently PixmapPacker is given each glyph with a unique name. This means if a TextureAtlas is updated, it'll have a ton (often ~100+) of regions that are just wasted space. PixmapPacker `pack` should accept null for a name and then not add those rects to the TextureAtlas.